packages/flutter/lib/src/rendering/paragraph.dart

```diff
@@ -1473,6 +1485,214 @@ class _SelectableFragment with Selectable, ChangeNotifier implements TextLayoutM
+          // Move the start edge to the closest word boundary.
+          targetPosition = _closestWordBoundary(wordBoundary, position);
+        }
+      }
+    } else {
+      // The position is not contained within the current rect. The targetPosition
+      // will either be at the end or beginning of the current rect. See [SelectionUtils.adjustDragOffset]
+      // for a more in depth explanation on this adjustment.
+      if (_selectableContainsOriginWord && existingSelectionStart != null && existingSelectionEnd != null) {
+        // When the selection is inverted by the new position it is necessary to
+        // swap the start edge (moving edge) with the end edge (static edge) to
+        // maintain the origin word within the selection.
+        final bool shouldSwapEdgesWhenSelectionIsNormalized = position.offset > existingSelectionEnd.offset;
+        final bool shouldSwapEdgesWhenSelectionNotNormalized = position.offset < existingSelectionEnd.offset;
+
+        if (existingSelectionStart.offset < existingSelectionEnd.offset) {
+          if (shouldSwapEdgesWhenSelectionIsNormalized || position.offset == existingSelectionEnd.offset) {
+            final ({TextPosition wordStart, TextPosition wordEnd}) localWordBoundary = _getWordBoundaryAtPosition(existingSelectionEnd);
+            assert(localWordBoundary.wordStart.offset >= range.start && localWordBoundary.wordEnd.offset <= range.end);
+            _setSelectionPosition(localWordBoundary.wordStart, isEnd: true);
+          }
+        } else {
+          if (shouldSwapEdgesWhenSelectionNotNormalized || position.offset == existingSelectionEnd.offset) {
+            final ({TextPosition wordStart, TextPosition wordEnd}) localWordBoundary = _getWordBoundaryAtPosition(existingSelectionEnd);
+            assert(localWordBoundary.wordStart.offset >= range.start && localWordBoundary.wordEnd.offset <= range.end);
+            _setSelectionPosition(localWordBoundary.wordEnd, isEnd: true);
+          }
+        }
+      }
+    }
+    return targetPosition ?? position;
+  }
+
+  TextPosition _updateSelectionEndEdgeByWord(
+    ({TextPosition wordStart, TextPosition wordEnd})? wordBoundary,
+    TextPosition position,
+    TextPosition? existingSelectionStart,
+    TextPosition? existingSelectionEnd,
+  ) {
+    TextPosition? targetPosition;
+    if (wordBoundary != null) {
+      assert(wordBoundary.wordStart.offset >= range.start && wordBoundary.wordEnd.offset <= range.end);
+      if (_selectableContainsOriginWord && existingSelectionStart != null && existingSelectionEnd != null) {
+        final bool shouldSwapEdgesWhenSelectionIsNormalized = position.offset < existingSelectionStart.offset;
+        final bool shouldSwapEdgesWhenSelectionNotNormalized = position.offset > existingSelectionStart.offset;
+        final bool shouldSwapEdges = existingSelectionStart.offset < existingSelectionEnd.offset ? shouldSwapEdgesWhenSelectionIsNormalized : shouldSwapEdgesWhenSelectionNotNormalized;
```

Comment:
```suggestion
        final bool isSamePosition = position.offset == existingSelectionStart.offset;
        final bool isSelectionInverted = existingSelectionStart.offset > existingSelectionEnd.offset;
        final bool shouldSwapEdges = !isSamePosition && (isSelectionInverted != (position.offset > existingSelectionStart.offset))
```

Reply:
Can you exit early if position is the same as before? If so this can be further simplified

Reply:
Thanks for the simplification! 

I think the case should be

```dart
final bool shouldSwapEdges = !isSamePosition && (isSelectionInverted != (position.offset < existingSelectionStart.offset));
```

When we are moving the end edge, If the selection is inverted, then it should only swap edges if the new position is after the start edge.

```
isSelectionInverted != position.offset < existingSelectionStart.offset
true != true // false, don't swap
true != false // true, swap
false != false // false, don't swap
false != true // true, swap
```

Reply:
What do you mean by exit early? As in exit `_updateSelectionEndEdgeByWord` early? If `position.offset == existingSelectionStart.offset` then we should keep `existingSelectionStart` in place and not change it.

Reply:
I meant exit _updateSelectionEndEdgeByWord early, Should we check it at the very beginning of the _updateSelectionEndEdgeByWord and just return since there isn't anything to do?

Reply:
Looking at the cases this might not be possible.

1. We are updating the end edge by word, but do not have a start edge. In this case a check at the beginning would return false and it would continue through the logic. This is the correct behavior.

2. We are updating the end edge by word, and we do have a start edge, but the selectable does not contain the origin word. In this case a check at the beginning would return immediately with `existingSelectionEnd`. The start edge on a selectable that doesn't contain the origin drag will either be at the start or end of the selectable. This isn't the expected behavior because it will keep the first or last word selected of a selectable that did not contain the drag origin.

3. We are updating the end edge by word, and we do have a start edge, and the selectable contains the origin word. In this case a check at the beginning would return immediately with `existingSelectionEnd` to keep the origin word in bounds. This is the expected behavior.

4. We are updating the end edge by word, the selectable does not contain the origin word, and the new position is not contained within the current rect. In this case a check at the beginning would result in `existingSelectionEnd ?? position`, if this is the first time visiting this selectable then it will most likely be `position` because the end edge has not been set yet, and if this is not the first time then it would be `existingSelectionEnd`. This is the expected behavior.

5. We are updating the end edge by word, the selectable contains the origin word, and the new position is not contained within the current rect. In this case a check at the beginning would skip swapping edges on `position.offset == existingSelectionStart.offset`. This is not the expected behavior.